Add Localisation support and file cacheing

TheBestAstroNOT opened the pull request:
PLEASE NOTE: For new files added via event editing. You MUST create a dummy BMD/BF uasset and add it via UnrealEssentials in order to modify it using Unreal Atlus Script.

TheBestAstroNOT committed:
Added localisation support

TheBestAstroNOT committed:
Merge branch 'main' of https://github.com/RyoTune/Unreal.AtlusScript

TheBestAstroNOT committed:
Added localisation support to the mod

NOTE: All existing assets will now be considered english assets. Another priority related note, if two mods add the same assets but if one adds it to english and the other adds it to another language, and the game ends up using the other language then the other language gets priority over english but if there is no asset for the other language but an english asset does exist then the english asset will be loaded.

TheBestAstroNOT committed:
Some optimisations

TheBestAstroNOT committed:
Corrected an oversight with the language detection optimisation

TheBestAstroNOT committed:
Fixed some more issues with the optimisation

TheBestAstroNOT committed:
Update AtlusAssetsRegistry.cs

TheBestAstroNOT committed:
More optimisations

TheBestAstroNOT committed:
Merge branch 'RyoTune:main' into main

TheBestAstroNOT committed:
random stuff

TheBestAstroNOT committed:
Update Atlus-Script-Tools

TheBestAstroNOT committed:
Update Atlus-Script-Tools

TheBestAstroNOT committed:
Better Asset Localisation Support and fixed compilation

Fixed asset compilation and added another folder called modern-altus-script that does not use english as a fallback language. Folder name can be easily renamed as it's a constant.

TheBestAstroNOT committed:
Bug Fixes

TheBestAstroNOT committed:
Removed modern-atlus-script folder and made all language specific files non-universal and files at the base root universal

TheBestAstroNOT committed:
Astrea base directory asset registration fix

TheBestAstroNOT committed:
Fix https://github.com/RyoTune/Unreal.AtlusScript/issues/1

TheBestAstroNOT committed:
Implement a file cache alongisde a new UNIVERSAL language option in the ESystemLanguage enum

TheBestAstroNOT committed:
Merge branch 'main' of https://github.com/TheBestAstroNOT/Unreal.AtlusScript

TheBestAstroNOT committed:
Add file check while loading cache

TheBestAstroNOT committed:
Small Bug Fix

TheBestAstroNOT committed:
Update ModConfig.json

TheBestAstroNOT committed:
Fixed issue where assets added via code are registered to english instead of universally, breaking CF and potentially other mods.

TheBestAstroNOT committed:
Fixed asset retrieval

TheBestAstroNOT committed:
Fix asset registration AGAIN

TheBestAstroNOT committed:
Improve modloading, move CacheRegistry inside the CacheFolder, move type AssetsMod to the IAtlusAssets and add a function to register an assets folder with support for caching.

TheBestAstroNOT committed:
Makes RegisterAssetsFolder get the asset language from the file path instead of an argument

Makes it so you don't have to have mutliple folders for each language with their own API call

TheBestAstroNOT committed:
Remove temp comments

TheBestAstroNOT committed:
Update AtlusAssetsRegistry.cs

RyoTune commented on Unreal.AtlusScript.Interfaces/IAtlusAssets.cs:
```diff
@@ -35,6 +67,12 @@ public interface IAtlusAssets
     void AddAsset(string name, string content, AssetType type, AssetMode mode);
 }
 
+public record AssetsMod(string ModId, string ModDir)
```
Moving this type to the interfaces library is unnecessary, `ModId` would never change and `ModDir` would be misnamed when used through the API meant for adding arbitrary folders to load from.

RyoTune commented on Unreal.AtlusScript.Interfaces/IAtlusAssets.cs:
```diff
@@ -47,3 +85,21 @@ public enum AssetMode
     Both,
 }
+
+public enum ESystemLanguage : byte
```
The first and default value should be `Universal`, since that's the default setting for assets. Not all caps, since it's not an acronym.

RyoTune commented on Unreal.AtlusScript.Reloaded/AtlusScript/AtlusAssetsRegistry.cs:
```diff
@@ -8,37 +12,136 @@ namespace Unreal.AtlusScript.Reloaded.AtlusScript;
-        }
-
-        foreach (var file in Directory.EnumerateFiles(mod.BaseAssetsDir, "*.*", SearchOption.AllDirectories))
+        fileCache.TryGetCacheByModID(mod.ModId.ToLower(), out var cacheContent);
```
`TryGetCacheByModID` should be responsible for formatting any arguments to whatever format it needs, not the calling function. You shouldn't have to "know" that it expects it to be lowercase, because someone else using it later might not.

RyoTune commented on Unreal.AtlusScript.Reloaded/AtlusScript/AtlusAssetsRegistry.cs:
```diff
@@ -8,37 +12,136 @@ namespace Unreal.AtlusScript.Reloaded.AtlusScript;
-
-        foreach (var file in Directory.EnumerateFiles(mod.BaseAssetsDir, "*.*", SearchOption.AllDirectories))
+        fileCache.TryGetCacheByModID(mod.ModId.ToLower(), out var cacheContent);
+        if (Directory.Exists(mod.BaseAssetsDir))
```
This section looks like a lot of code was copy and pasted when it probably didn't need to be, keyword being *probably* since I haven't looked over it completely.

RyoTune commented on Unreal.AtlusScript.Reloaded/AtlusScript/AtlusAssetsRegistry.cs:
```diff
@@ -8,37 +12,136 @@ namespace Unreal.AtlusScript.Reloaded.AtlusScript;
+                        ESystemLanguage dirLang = GetFileLang(mod.AstreaAssetsDir, dir);
+                        foreach (var file in Directory.EnumerateFiles(dir, "*.*", SearchOption.AllDirectories))
+                        {
+                            var identifier = new FileIdentifier(Path.GetFileName(file), dirLang, AssetMode.Astrea);
```
This code section absolutely is though and the repeated code should just be its own function.

RyoTune commented on Unreal.AtlusScript.Reloaded/AtlusScript/AtlusAssetsRegistry.cs:
```diff
@@ -50,38 +153,173 @@ public bool TryGetAsset(AssetMode mode, string assetName, [NotNullWhen(true)]out
+        {
+            this.AddAssetFile(file, mode, ESystemLanguage.UNIVERSAL);
+        }
+    }
```
These functions are unnecessarily duplicated with the original deprecated, seemingly just so they match the new `Register` naming scheme? The new functions should instead be matching the original's `Add` naming, to not confuse existing uses of the API.

RyoTune commented on Unreal.AtlusScript.Reloaded/AtlusScript/AtlusAssetsRegistry.cs:
```diff
@@ -50,38 +153,173 @@ public bool TryGetAsset(AssetMode mode, string assetName, [NotNullWhen(true)]out
+                    assetMode = mode,
+                    hashedName = hashedName
+                });
+                string filePath = Path.Combine(fileCache.CacheFolder, modData.ModId.ToLower(), (currentAssetLang.ToString() != "UNIVERSAL") ? currentAssetLang.ToString().ToLower() : string.Empty, mode.ToString().ToLower(), hashedName);
```
- The logic for creating the output file path is too complex for a single line, as well being duplicated. It should be in its own function.
- Also, use `Path.Join` instead of `Path.Combine` (google why).
- You shouldn't compare enums to their string value for comparisons: `currentAssetLang.ToString() != "UNIVERSAL"`, just compare against the enum value itself `Lang.UNIVERSAL`. This code will break when you rename `UNIVERSAL` as is.

RyoTune commented on Unreal.AtlusScript.Reloaded/AtlusScript/AtlusAssetsRegistry.cs:
```diff
@@ -50,38 +153,173 @@ public bool TryGetAsset(AssetMode mode, string assetName, [NotNullWhen(true)]out
+                File.WriteAllBytes(filePath, msgAsset.Data);
+                fileCache.Save();
+            }
+            Log.Debug($"Registered MSG ({mode}): {msgAsset.Name}, language: {currentAssetLang}");
```
This is just my personal style for mod logging, but you should separate variables with `||` or `/`, with title capitalization: `$"Registered MSG ({mode}): {msgAsset.Name} / Language: {currentAssetLang}"`

RyoTune commented on Unreal.AtlusScript.Reloaded/AtlusScript/AtlusAssetsRegistry.cs:
```diff
@@ -50,38 +153,173 @@ public bool TryGetAsset(AssetMode mode, string assetName, [NotNullWhen(true)]out
+            this.assetContainers[currentAssetLang].Add(flowAsset);
+            if (modData != null && flowAsset.Data != null)
+            {
+                byte[] hashBytes = sha1.ComputeHash(Encoding.UTF8.GetBytes(Path.GetFileName(file)));
```
What's the point of taking the hash here instead of just the file name? If multiple mods replace the same file, they'll replace the same file in the cache too. If that's the point then the file name itself could be used instead of the hash?

RyoTune commented on Unreal.AtlusScript.Reloaded/AtlusScript/AtlusAssetsRegistry.cs:
```diff
@@ -50,38 +153,173 @@ public bool TryGetAsset(AssetMode mode, string assetName, [NotNullWhen(true)]out
+        }
+        else
+        {
+            Log.Debug($"Failed to load asset from Cache ({identifier.AssetMode}): {identifier.Name}, language: {identifier.Language.ToString()}");
```
`ToString` is unnecessary here; when using string interpolation, it'll use it automatically when passed non-string variables.

RyoTune commented on Unreal.AtlusScript.Reloaded/AtlusScript/AtlusScriptService.cs:
```diff
@@ -26,27 +26,44 @@ internal unsafe class AtlusScriptService
+    }
 
-        _GetLanguage = new SHFunction<GetLanguage>("48 89 5C 24 ?? 48 89 6C 24 ?? 48 89 74 24 ?? 57 41 56 41 57 48 83 EC 30 E8");
+    private ESystemLanguage GetGameLanguage()
```
Because this calls a native function, it'll either work or crash, there is no `try` :c
`gameLanguage` should be made non-nullable (after making `Universal` the first value in the enum) and the `try/catch` removed.

RyoTune commented on Unreal.AtlusScript.Reloaded/AtlusScript/AtlusScriptService.cs:
```diff
@@ -105,18 +122,23 @@ private void OnObjectCreated(UnrealObject obj)
 
         var mode = this.game.IsAstrea() ? AssetMode.Astrea : AssetMode.Default;
-        if (this.assetsRegistry.TryGetAsset(mode, obj.Name, out var data))
+        if(gameLanguage==null)
```
Compare against `Universal` since it's no longer nullable. Probably "nicer" to make a `hasLangLoaded` `bool`, but either works.

RyoTune commented on Unreal.AtlusScript.Reloaded/AtlusScript/AtlusScriptService.cs:
```diff
@@ -105,18 +122,23 @@ private void OnObjectCreated(UnrealObject obj)
+        }
+        if (this.assetsRegistry.TryGetAsset(mode, obj.Name, out var data, (ESystemLanguage)gameLanguage!))
+        {
+                var objAsset = (UAtlusScriptAsset*)obj.Self;
```
Style nitpick, but please try maintain the project formatting and indentation.

RyoTune commented on Unreal.AtlusScript.Reloaded/Config.cs:
```diff
@@ -39,6 +44,24 @@ public enum Decomp_Endianess
     LITTLE_ENDIAN,
 }
 
+public enum AssetConfigLanguage
```
Why is there a duplicate language enum?

RyoTune commented on Unreal.AtlusScript.Reloaded/Mod.cs:
```diff
@@ -64,8 +65,49 @@ public Mod(ModContext context)
-        this.atlusScript = new(uobjects!, unreal!, this.atlusRegistry, flowDecompiler, gameLibrary, modDir);
+
+        //Generate/Load the file cache registry
+        if (!Directory.Exists(Path.Combine(modDir, "Cache")))
```
Way too much logic here for a constructor, and too complex for a function in `Mod.cs`. This code should be in whatever service is responsible for caching (currently AtlusAssetsRegistry), but that's already a bit bloated. Caching should be moved to it's own service which the registry then uses.

RyoTune commented:
> @RyoTune would love it if u could test it on ur system too, but it seems to work on my system just fine. It's reduced load times with the femc mod from an average of 70,000-80,000 ms to around 10,000 ms during testing on a mid range laptop.

Did you test this thoroughly? Specifically, running it once to build cache, deleting it, running again for a base line, and running again *right* after for the cached speed.

It's important to have .NET stuff run once before measuring speed because it only fully compiles code while it's running, so .NET stuff generally runs faster the first run. I'll be testing it later myself.

TheBestAstroNOT commented:
oh im so sorry, i took the a bit of the caching system from PEssentials though most of the functionality is implemented by me. It was so long b/w when i started work on it and me getting to actually test it that I totally forgot to credit them. Also thanks to Sewer56 for some advice with the caching system.

TheBestAstroNOT commented:
> > @RyoTune would love it if u could test it on ur system too, but it seems to work on my system just fine. It's reduced load times with the femc mod from an average of 70,000-80,000 ms to around 10,000 ms during testing on a mid range laptop.
> 
> Did you test this thoroughly? Specifically, running it once to build cache, deleting it, running again for a base line, and running again _right_ after for the cached speed.
> 
> It's important to have .NET stuff run once before measuring speed because it only fully compiles code while it's running, so .NET stuff generally runs faster the first run. I'll be testing it later myself.
> 
Also, multiple people have tested it and I myself have tested it repeatedly by deleting the cache multiple times.

TheBestAstroNOT committed:
Add suggested changes

I haven't touched the repetitive code in RegisterMod because it's functional and I can't really test it anymore if I modify it.

RyoTune committed:
Adjust function names.
